Realm/ObjectServerTests/SwiftCollectionSyncTests.swift

```diff
@@ -338,3 +338,602 @@ class CollectionSyncTestCase: SwiftSyncTestCase {
                                                   .objectId(ObjectId("507f1f77bcf86cd799439011"))])
     }
 }
+
+@available(macOS 13, *)
+class AsyncAnyRealmValueSyncTest: SwiftSyncTestCase {
+    override class var defaultTestSuite: XCTestSuite {
+        // async/await is currently incompatible with thread sanitizer and will
+        // produce many false positives
+        // https://bugs.swift.org/browse/SR-15444
+        if RLMThreadSanitizerEnabled() {
+            return XCTestSuite(name: "\(type(of: self))")
+        }
+        return super.defaultTestSuite
+    }
+
+    override func configuration(user: User) -> Realm.Configuration {
+        user.flexibleSyncConfiguration(initialSubscriptions: {
+            $0.append(QuerySubscription<SwiftTypesSyncObject>())
+            $0.append(QuerySubscription<SwiftPerson>())
+        })
+    }
+
+    override var objectTypes: [ObjectBase.Type] {
+        [SwiftTypesSyncObject.self, SwiftPerson.self]
+    }
+
+    override func createApp() throws -> String {
+        try createFlexibleSyncApp()
+    }
+
+    private var sampleData: Array<AnyRealmValue> {
+        let so = SwiftPerson()
+        so.firstName = name
+        let oid = ObjectId.generate()
+        let uuid = UUID()
+        let date = Date()
+        return [
+            .string("hello"),
+            .bool(false),
+            .int(234),
+            .double(12345.678901),
+            .float(12.34),
+            .data(Data("a".utf8)),
+            .date(date),
+            .object(so),
+            .objectId(oid),
+            .uuid(uuid),
+            .decimal128(Decimal128(number: 567))
+        ]
+    }
+
+    private func assertListEqual(_ object: SwiftTypesSyncObject, _ index: Int, _ value: AnyRealmValue) {
+        if case .object = value {
+            XCTAssertEqual(object.anyCol.listValue?[index].object(SwiftPerson.self)?.firstName, value.object(SwiftPerson.self)?.firstName)
+        } else if case .float = value {
+            if object.anyCol.listValue?[index].floatValue != nil {
+                XCTAssertEqual(object.anyCol.listValue?[index], value)
+            } else {
+                // This happens when we download a fresh realm
+                XCTAssertEqual(object.anyCol.listValue?[index], .double(Double(value.floatValue!)))
+            }
+        } else if case .date = value {
+            // Date is not exact when synced
+            let date = object.anyCol.listValue?[index].dateValue
+            XCTAssertTrue(Calendar.current.isDate(date!, equalTo: value.dateValue!, toGranularity: .second))
+        } else {
+            XCTAssertEqual(object.anyCol.listValue?[index], value)
+        }
```

Comment:
All of the checking logic here is duplicated three times and it should be pretty easy to extract.